dask/array/core.py

```diff
@@ -232,46 +233,90 @@ def slices_from_chunks(chunks):
-    {('X', 0, 0): (getter, 'X', (slice(0, 2), slice(0, 3))),
-     ('X', 1, 0): (getter, 'X', (slice(2, 4), slice(0, 3))),
-     ('X', 1, 1): (getter, 'X', (slice(2, 4), slice(3, 6))),
-     ('X', 0, 1): (getter, 'X', (slice(0, 2), slice(3, 6)))}
+    inline_array=True,
+) -> HighLevelGraph:
+    """
+    HighLevelGraph for slicing chunks from an array-like according to a chunk pattern.
+
+    If ``inline_array`` is True, this make a Blockwise layer of slicing tasks where the
+    array-like is embedded into every task.,
+
+    If ``inline_array`` is False, this inserts the array-like as a standalone value in
+    a MaterializedLayer, then generates a Blockwise layer of slicing tasks that refer
+    to it.
+
+    >>> dict(graph_from_arraylike(arr, chunks=(2, 3), shape=(4, 6), name="X", inline_array=True))  # doctest: +SKIP
+    {(arr, 0, 0): (getter, arr, (slice(0, 2), slice(0, 3))),
+     (arr, 1, 0): (getter, arr, (slice(2, 4), slice(0, 3))),
+     (arr, 1, 1): (getter, arr, (slice(2, 4), slice(3, 6))),
+     (arr, 0, 1): (getter, arr, (slice(0, 2), slice(3, 6)))}
+
+    >>> dict(
+            graph_from_arraylike(
+                arr, chunks=((2, 2), (3, 3)), shape=(4,6), name="X", inline_array=False
+            )
+    )  # doctest: +SKIP
+    {"original-X": arr,
+     ('X', 0, 0): (getter, 'original-X', (slice(0, 2), slice(0, 3))),
+     ('X', 1, 0): (getter, 'original-X', (slice(2, 4), slice(0, 3))),
+     ('X', 1, 1): (getter, 'original-X', (slice(2, 4), slice(3, 6))),
+     ('X', 0, 1): (getter, 'original-X', (slice(0, 2), slice(3, 6)))}
     """
-    out_name = out_name or arr
     chunks = normalize_chunks(chunks, shape, dtype=dtype)
-    keys = product([out_name], *(range(len(bds)) for bds in chunks))
-    slices = slices_from_chunks(chunks)
+    out_ind = tuple(range(len(shape)))
 
     if (
         has_keyword(getitem, "asarray")
         and has_keyword(getitem, "lock")
         and (not asarray or lock)
     ):
-        values = [(getitem, arr, x, asarray, lock) for x in slices]
+        getter = partial(getitem, asarray=asarray, lock=lock)
     else:
         # Common case, drop extra parameters
-        values = [(getitem, arr, x) for x in slices]
+        getter = getitem
 
-    return dict(zip(keys, values))
+    if inline_array:
+        layer = core_blockwise(
+            getter,
+            name,
+            out_ind,
+            arr,
+            None,
+            ArraySliceDep(chunks),
+            out_ind,
+            numblocks={},
+        )
+        return HighLevelGraph.from_collections(name, layer)
+    else:
+        original_name = "original-" + name
+
+        layers = {}
+        layers[original_name] = MaterializedLayer({original_name: arr})
+        layers[name] = core_blockwise(
+            getter,
+            name,
+            out_ind,
+            original_name,
+            None,
+            ArraySliceDep(chunks),
+            out_ind,
+            numblocks={},
+        )
```

Comment:
nit: For readability, it might be nice to not repeat this `core_blockwise` call twice in the if/else cases, but just have it once so it's clearer to readers what the difference is/isn't between the two cases?
```python
layers = {}
deps = {out_name: set()}
if not inline_array:
    original_name = "original-" + name
    layers[original_name] = MaterializedLayer({original_name: arr})
    deps[out_name].add(original_name)
    arr = original_name

layers[name] = core_blockwise(
    getter,
    name,
    out_ind,
    original_name,
    None,
    ArraySliceDep(chunks),
    out_ind,
    numblocks={},
)
return HighLevelGraph(layers, deps)
```

Reply:
I actually experimented with this approach a few days ago, and came to the opposite conclusion -- that having the repeated `core_blockwise` invocation was not too bad in exchange for making it easier to see the difference between the two code paths.

That is to say: in exchange for having a single `core_blockwise` you have to spend more time jumping up and down the block, thinking about whether `arr` is a real array or whether it has been changed to a string, etc. At the end of the day, I thought it was simpler to have the two paths more separate and linear.